File: LeetCode/20. Valid Parentheses/elixir/solution.exs
defmodule Solution do
  @spec is_valid(s :: String.t()) :: boolean
  def is_valid(s) do
    s
    |> to_charlist()
    |> valid_parentheses()
  end

  defp valid_parentheses(chars, stack \\ [])

  defp valid_parentheses([], stack) do
    stack == []
  end

  defp valid_parentheses([h | t], []) do
    valid_parentheses(t, [h])
  end

  defp valid_parentheses([h | t], [top | rest] = stack) do
    if is_complete_brackets(h, top) do
      valid_parentheses(t, rest)
    else
      valid_parentheses(t, [h | stack])
    end
  end

  defp is_complete_brackets(a, b) do
    case {b, a} do
      {?(, ?)} -> true
      {?{, ?}} -> true
      {?[, ?]} -> true
      _ -> false
    end
  end
end

File: LeetCode/20. Valid Parentheses/java/Solution.java
class Solution {
    public boolean isValid(String s) {
        Stack<Character> stack = new Stack<>();
        for (char c : s.toCharArray()) {
            if (!stack.isEmpty() && isCorrectBracket(stack.peek(), c)) {
                stack.pop();
            } else {
                stack.push(c);
            }
        }
        return stack.isEmpty();
    }

    private boolean isCorrectBracket(char a, char b) {
        return a == '(' && b == ')' || a == '[' && b == ']' || a == '{' && b == '}';
    }
}
